File: sdk/go/conformance/harness.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

// Package conformance provides the CTK harness contract, runner, and
// reference harness (§13.2).
//
// The runner (RunVector) and scripted interceptor/resolver evaluation
// delegate to the Rust core via agenthooks.Ctk*; this package keeps
// only the Harness protocol (native callback into the framework under
// test), a recording wrapper, and Scenario helpers. See
// conformance/RUNNER.md for the shape every language SDK follows.
package conformance

import (
	"context"

	"github.com/responsibleai/agent-hooks/sdk/go/agenthooks"
)

// Capability is a host-declared capability (§3.2).
type Capability string

const (
	ModelCalls        Capability = "model_calls"
	ToolCalls         Capability = "tool_calls"
	ParallelToolCalls Capability = "parallel_tool_calls"
	Streaming         Capability = "streaming"
	MultiTurn         Capability = "multi_turn"
	// Int64JSON marks a harness language that can hold >2^53 integers
	// from vector JSON losslessly (§4.4). JavaScript harnesses omit it.
	Int64JSON Capability = "int64_json"
	// BigintJSON: the harness JSON layer preserves integer tokens
	// beyond u64/i64 (Go: json.Number vector decoding).
	BigintJSON Capability = "bigint_json"
)

// RunOutcome describes how a harness run ended.
type RunOutcome string

const (
	Completed RunOutcome = "completed"
	Blocked   RunOutcome = "blocked"
	Suspended RunOutcome = "suspended"
	Errored   RunOutcome = "error"
)

// Scenario is a hermetic scripted run loaded from a CTK vector (wire-shaped).
type Scenario struct {
	Input       map[string]any   `json:"input"`
	Tools       []map[string]any `json:"tools"`
	ModelScript []map[string]any `json:"model_script"`
}

// ToolInvocation is one entry in the harness's mock-tool log.
type ToolInvocation struct {
	Name string         `json:"name"`
	Args map[string]any `json:"args"`
}

// IdentityPair is one (input_identity, enforced_identity) pair per
// interception, taken from the emitter's InterceptionRecords so the
// CTK can assert expect.identities_equal. Identities are nil when the
// identity provider is nil (§10.1).
type IdentityPair struct {
	InputIdentity    *string `json:"input_identity"`
	EnforcedIdentity *string `json:"enforced_identity"`
}

// RunRecord is what Harness.Run returns to the CTK runner.
type RunRecord struct {
	Outcome         RunOutcome
	FinalOutput     any
	ToolInvocations []ToolInvocation
	Err             string
	// Identities is one entry per interception, in order, from the
	// harness's emitter.
	Identities []IdentityPair
	// Records is the wire-shaped InterceptionRecords (§10.3), one per
	// emission, in order. Enables expect.records assertions.
	Records []agenthooks.InterceptionRecord
}

// Harness is the single interface a framework adapter implements for the CTK.
type Harness interface {
	Name() string
	Capabilities() map[Capability]struct{}

	// Setup wires the scenario's mock model + tools into the framework,
	// registers the interceptors and resolver, and sets the enforcement
	// mode, the vector's composition profile (§7.1), and its identity
	// provider (§10.1; nil = identity-unbound — vectors declare
	// "jcs-sha256" or null, custom providers are not vector-expressible).
	// When redactForApproval is non-empty the harness MUST register a §9
	// approval redactor that replaces each listed §5.2 path in the
	// request context's target with the string "[redacted]" (write-back
	// mirrored per §4.3), leaving unresolvable paths untouched.
	Setup(
		scenario Scenario,
		interceptors []agenthooks.Interceptor,
		resolver agenthooks.ApprovalResolver,
		mode agenthooks.EnforcementMode,
		composition agenthooks.CompositionConfig,
		identityProvider *agenthooks.IdentityProvider,
		redactForApproval []string,
	) error

	Run(ctx context.Context) (RunRecord, error)

	Teardown()
}

File: sdk/python/python/agent_hooks/ctk/harness.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
"""CTK harness contract a framework adapter implements once (§13.2)."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Protocol

from agent_hooks._types import EnforcementMode
from agent_hooks.approval import ApprovalResolver
from agent_hooks.composition import CompositionConfig
from agent_hooks.interceptor import Interceptor


class Capability(str, Enum):
    """Host-declared capability subset (§3.2)."""

    MODEL_CALLS = "model_calls"
    TOOL_CALLS = "tool_calls"
    PARALLEL_TOOL_CALLS = "parallel_tool_calls"
    STREAMING = "streaming"
    MULTI_TURN = "multi_turn"
    #: The harness language can hold >2^53 integers from vector JSON
    #: losslessly (§4.4). JavaScript harnesses omit this.
    INT64_JSON = "int64_json"
    BIGINT_JSON = "bigint_json"
    #: The host declares buffered_output: false and mediates
    #: post_model_call incrementally under the §12.1 exception with
    #: watermark-gated release; gates the streaming/incremental vector
    #: part. Buffered hosts (the default) omit this and skip it.
    INCREMENTAL_OUTPUT = "incremental_output"


class RunOutcome(str, Enum):
    COMPLETED = "completed"
    BLOCKED = "blocked"
    SUSPENDED = "suspended"
    ERROR = "error"


@dataclass(slots=True)
class ToolBehavior:
    when_args: dict[str, Any] | None
    return_: Any
    is_error: bool = False


@dataclass(slots=True)
class ToolSpec:
    name: str
    behavior: list[ToolBehavior]
    schema: dict[str, Any] = field(default_factory=dict)

    def invoke(self, args: dict[str, Any]) -> tuple[Any, bool]:
        """Mock-tool dispatch: first matching behavior wins."""
        for b in self.behavior:
            if b.when_args is None or b.when_args == args:
                return b.return_, b.is_error
        raise AssertionError(
            f"tool {self.name!r} invoked with {args!r}: no matching behavior clause"
        )


@dataclass(slots=True)
class ModelResponse:
    content: Any
    tool_calls: list[dict[str, Any]]
    finish_reason: str


@dataclass(slots=True)
class Scenario:
    """Hermetic scripted run loaded from a CTK vector."""

    input: dict[str, Any]
    tools: dict[str, ToolSpec] = field(default_factory=dict)
    model_script: list[ModelResponse] = field(default_factory=list)

    @classmethod
    def from_wire(cls, obj: dict[str, Any]) -> Scenario:
        tools = {
            t["name"]: ToolSpec(
                name=t["name"],
                schema=t.get("schema", {}),
                behavior=[
                    ToolBehavior(
                        when_args=b.get("when_args"),
                        return_=b["return"],
                        is_error=b.get("is_error", False),
                    )
                    for b in t["behavior"]
                ],
            )
            for t in obj.get("tools", [])
        }
        model_script = [
            ModelResponse(
                content=m["respond"]["content"],
                tool_calls=list(m["respond"]["tool_calls"]),
                finish_reason=m["respond"]["finish_reason"],
            )
            for m in obj.get("model_script", [])
        ]
        return cls(input=obj["input"], tools=tools, model_script=model_script)


@dataclass(slots=True)
class RunRecord:
    """What :meth:`Harness.run` returns to the CTK runner."""

    outcome: RunOutcome
    final_output: Any | None
    tool_invocations: list[dict[str, Any]] = field(default_factory=list)
    error: str | None = None
    #: ``(input_identity, enforced_identity)`` per interception, in order,
    #: from the harness's emitter (``None`` when the identity provider is
    #: ``null``, §10.1). Enables ``expect.identities_equal``.
    identities: list[tuple[str | None, str | None]] = field(default_factory=list)
    #: Wire-shaped ``InterceptionRecord`` dicts (§10.3), one per emission,
    #: in order. Enables ``expect.records`` assertions.
    records: list[dict[str, Any]] = field(default_factory=list)


class Harness(Protocol):
    """The single interface a framework adapter implements for the CTK."""

    name: str
    capabilities: set[Capability]

    def setup(
        self,
        scenario: Scenario,
        interceptors: list[Interceptor],
        resolver: ApprovalResolver | None,
        mode: EnforcementMode,
        composition: CompositionConfig,
        identity_provider: str | None,
        redact_for_approval: list[str] | None = None,
    ) -> None: ...

    async def run(self) -> RunRecord: ...

    def teardown(self) -> None: ...
